### ISSUE TITLE
Failure constructor API doesn't specify valid values of exc_tb

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2496]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#2496|
|Type|defect|
|Created|2007-02-28 07:58:18Z|
|Branch|https://github.com/twisted/twisted/tree/failure-exctb-2496|

`failure.Failure.__init__`'s docstring doesn't say what values are allowed for exc_tb. It suggests that traceback objects are allowed, but doesn't say that other things (such as strings) might be allowed.

Marking as high to vent some frustration.

<details><summary>Searchable metadata</summary>

```
trac-id__2496 2496
type__defect defect
reporter__jml jml
priority__high high
milestone__ 
branch__branches_failure_exctb_2496 branches/failure-exctb-2496
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1172649498000000 1172649498000000
changetime__1220229514000000 1220229514000000
version__None None
owner__ 
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

As a first approximation, I would say that _only_ traceback instances or None are allowed.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra commented|
|-|-|

(In [24561]) Branching to 'failure-exctb-2496'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Replying to [exarkun](#note_1):
> As a first approximation, I would say that _only_ traceback instances or None are allowed.

Added this note in r24562, putting it up for review. I'd like to see a actual description of ```exc_tb``` as well but can't come up with something useful, leaving it up to the twisted guru's.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

I'm not sure if strings were allowed for `exc_tb` at some point in the past, but the certainly aren't now.  The object must look a lot like a traceback object in order for `Failure.__init__` to complete successfully.  Additionally, a `Failure` which is used with `inlineCallbacks` must really be a traceback object, since it is passed to `GeneratorType.throw` which requires exactly the right type.

Oh hey, it just occurred to me that perhaps jml was frustrated because he had a `CopiedFailure` instead of a real `Failure`?  `CopiedFailure` is a terrible thing with a lot of badness around it, but that badness doesn't need to be allowed to pollute the comparably mild badness of `Failure` itself.

So yea, traceback or None is correct here.  I'd suggest changing the doc improvement to use epytext markup, though.  It could also explain that if `None` is supplied for `exc_value`, the value of `exc_tb` is ignored, otherwise if `exc_tb` is `None`, it will be found from execution context (ie, `sys.exc_info`).

Thanks

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra removed owner|
|-|-|

Updated the value for ```exc_tb``` in r24638, putting it up for review. And what about getting rid of those lines that are already commented out?

```python
#             else:
#                 log.msg("Erf, %r created with no traceback, %s %s." % (
#                     repr(self), repr(exc_value), repr(exc_type)))
#                 for s in traceback.format_stack():
#                     log.msg(s)
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

Changes from two spaces after a . to one space after a . aren't needed (two is a widely used convention - one is okay too in Twisted, but changing from one to the other is pointless).

Perhaps the `X{}` markup could be fixed here as well, though.  `X{}` isn't valid epytext as far as I can tell.  The documentation uses it as an example of the _form_ of markup.  Perhaps someone got confused.  I'd just delete it.

In the sentence _For C{exc_tb} only L{traceback} instances or C{None} is allowed._, use _are_ instead of _is_.

Otherwise looks good, please merge when the above points are addressed.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [24656]) Merge failure-exctb-2496: update docstring for exc_tb's value for Failure

Author: thijs
Reviewer: exarkun
Fixes: [#2496](https://github.com/twisted/twisted/issues/2496)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
